x/emissions/abci_test.go

```diff
@@ -1,229 +1,180 @@
+	totalRewardCoins := sdk.NewCoins(sdk.NewCoin(config.BaseDenom, totalBlockRewards.TruncateInt()))
+	assert.NoError(t, err)
+	// Fund the emission pool to start the emission process
+	err = sk.BankKeeper.MintCoins(ctx, emissionstypes.ModuleName, totalRewardCoins)
+	assert.NoError(t, err)
+
+	// Setup module accounts for emission pools
+	undistributedObserverPoolAddress := sk.AuthKeeper.GetModuleAccount(ctx, emissionstypes.UndistributedObserverRewardsPool).GetAddress()
+	undistributedTssPoolAddress := sk.AuthKeeper.GetModuleAccount(ctx, emissionstypes.UndistributedTssRewardsPool).GetAddress()
+	feeCollecterAddress := sk.AuthKeeper.GetModuleAccount(ctx, types.FeeCollectorName).GetAddress()
+	emissionPool := sk.AuthKeeper.GetModuleAccount(ctx, emissionstypes.ModuleName).GetAddress()
+
+	blockRewards := emissionstypes.BlockReward
+	observerRewardsForABlock := blockRewards.Mul(sdk.MustNewDecFromStr(k.GetParams(ctx).ObserverEmissionPercentage)).TruncateInt()
+	validatorRewardsForABlock := blockRewards.Mul(sdk.MustNewDecFromStr(k.GetParams(ctx).ValidatorEmissionPercentage)).TruncateInt()
+	tssSignerRewardsForABlock := blockRewards.Mul(sdk.MustNewDecFromStr(k.GetParams(ctx).TssSignerEmissionPercentage)).TruncateInt()
+	distributedRewards := observerRewardsForABlock.Add(validatorRewardsForABlock).Add(tssSignerRewardsForABlock)
+	assert.True(t, blockRewards.TruncateInt().GT(distributedRewards))
+
+	for i := 0; i < 100; i++ {
+		emissionPoolBeforeBlockDistribution := sk.BankKeeper.GetBalance(ctx, emissionPool, config.BaseDenom).Amount
+		// produce a block
+		emissionsModule.BeginBlocker(ctx, *k)
+
+		// Assert distribution amount
+		emissionPoolBalanceAfterBlockDistribution := sk.BankKeeper.GetBalance(ctx, emissionPool, config.BaseDenom).Amount
+		assert.True(t, emissionPoolBeforeBlockDistribution.Sub(emissionPoolBalanceAfterBlockDistribution).Equal(distributedRewards))
+
+		// totalDistributedTillCurrentBlock is the net amount of rewards distributed till the current block, this works in a unit test as the fees are not being collected by validators
+		totalDistributedTillCurrentBlock := sk.BankKeeper.GetBalance(ctx, feeCollecterAddress, config.BaseDenom).Amount.
+			Add(sk.BankKeeper.GetBalance(ctx, undistributedObserverPoolAddress, config.BaseDenom).Amount).
+			Add(sk.BankKeeper.GetBalance(ctx, undistributedTssPoolAddress, config.BaseDenom).Amount)
+		// Assert we are always under the max limit of block rewards
+		assert.True(t, totalRewardCoins.AmountOf(config.BaseDenom).
+			Sub(totalDistributedTillCurrentBlock).GTE(sdk.ZeroInt()))
+
+		ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+	}
+
+	// We can simplify the calculation as the rewards are distributed equally among all the observers
+	rewardPerUnit := observerRewardsForABlock.Quo(sdk.NewInt(int64(len(ballotList) * len(observerSet.ObserverList))))
+	emissionAmount := rewardPerUnit.Mul(sdk.NewInt(int64(len(ballotList))))
```

Comment:
The problem is we replicate logic of the function to test in the test which assume that this logic is correct.

It would be beter to directly define and check [input/expected output] with simple values we know are correct and check with actual output.

I think the best structure for tests for a distribution logic would be to use the Table Driven Test paradigm:
https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Reply:
This test tries to ensure that the observers receive the balance they should have . The distribution
logic itself already has some table-driven tests for the `BuildRewardsDistribution` function.

With that said, I see some value in adding tests to calculate the reward amount ,from the distribution map . I would not want to replace this test , and add that in separately

Reply:
>The distribution
logic itself already has some table-driven tests for the BuildRewardsDistribution function.

What do you mean by this? Any link?

Reply:
I was referring to this test 
https://github.com/zeta-chain/zeta-node/blob/0c9ca3df5d30a6ca4b661eeac5fcd000e45d6e9c/x/observer/types/ballot_test.go#L347-L347